### DISCUSSION TITLE
Escaping characters with $like and $ilike for postgresql.

### DISCUSSION THREAD

**Discussion by Weplik** [mikro-orm/mikro-orm#4380] [General]
**Is your feature request related to a problem? Please describe.**
Escaping characters when using `$like` and `$ilike` for postgresql. Now I need to write additional functions to support this operation, and this is very inconvenient, since you can forget about it in some place. 

**Describe the solution you'd like**
When I use `$like` or `$ilike`, then I want escaping characters to happen. 

Example:
```
Code: repository.find({ title: { $like: '%_%' } })

Current SQL: SELECT ... WHERE title like '%_%'

Expected SQL: SELECT ... WHERE title like '%\_%'
```

**Describe alternatives you've considered**
1. In the client configuration there would be an option `escapeCharactersInPatternMatching?: boolean`. If `true`, then escaping characters is performed by default.
2. In the `find`, `findOne` and `findAndCount` methods, add `escapeCharactersInPatternMatching?: boolean` to the options.

**Additional context**
1. https://stackoverflow.com/questions/38084864/how-to-escape-underscores-in-postgresql
2. https://www.postgresql.org/docs/current/functions-matching.html

**Comment by B4nan** [comment]
I don't think this is a good idea, `_` is a special character with a meaning, we can't just escape it - you might want to actually use it.

**Reply by Weplik** [reply]
> I don't think this is a good idea, `_` is a special character with a meaning, we can't just escape it - you might want to actually use it.

I totally agree that this should not be the default behavior.

Therefore, I suggested adding an option `escapeCharactersInPatternMatching` so that the developer could make a decision himself depending on the business logic. For example, in my application it is necessary to escape characters everywhere. And it seems to me that there is no need to escape characters, only in some more technical applications. And when the application is for ordinary users and they enter `_` in the search, it means that a match with `_` is needed.

**Reply by B4nan** [reply]
The problem is that `_` is a special character just like `%`, I don't understand why would you expect some magic escaping that only works on `_` but ignores `%`, that seems just wrong and super confusing.

**Reply by Weplik** [reply]
Okay, imagine that there is a project entity in the system, it has a `title` field. User wants to find all projects with a `title` that starts with `my_` (I think this is a frequent case in most systems). And here, if we don't escape the characters, then we will run into how the database works, and in all such places we need to write some additional functions.

And my suggestion was to abandon additional functions, and dictate such behavior through a new option.

> that seems just wrong and super confusing

It only looks wrong when ORM would always escape the characters. And this behavior is set by the developer.

**Reply by B4nan** [reply]
I understand what you want, but you are clearly missing my point. `_` is a wildcard just like `%`, you can't treat those two differently.

**Reply by Weplik** [reply]
Okay, how do you see then that there is an option `escapeCharactersInPatternMatching?: string[]`, where the developer can specify escaped characters? There are more opportunities to implement any logic at the ORM level.

**Reply by B4nan** [reply]
There are only two special characters for like queries, exactly those two we are talking about `_` and `%`. 

I really don't think it makes sense to have some weird partial escaping option, this needs to be done by you - the developer, who actually knows what part of that string should be escaped and what not, this depends on case by case. 

Back to your original case, you were arguing about a user trying to search for `_` character - do you understand that the user can also search for `%`? It's the same thing, you need to escape both, but not the outer `%` chars. This is something specific for your app, not a general problem, one could do a like search where they want exact match, so escaping all the `%` chars would be needed. 

So nope, I don't see any reason to add something like that. If you want to reduce duplication from your app, just define a function that does the escaping for you.

**Reply by Weplik** [reply]
Okay, I get your point.

And can you tell me if there is or is planned to make a method for escaping the characters in ORM?
